phoenix-core/src/main/java/org/apache/phoenix/index/IndexMaintainer.java

```diff
@@ -694,16 +696,22 @@ private Void addDataColInfo(final PTable dataTable, Expression expression) {
                         output.write(ptr.get(), ptr.getOffset(), ptr.getLength());
                     }
                 }
+
                 if (!indexColumnType.isFixedWidth()) {
-                    output.writeByte(SchemaUtil.getSeparatorByte(rowKeyOrderOptimizable, ptr.getLength() == 0, isIndexColumnDesc ? SortOrder.DESC : SortOrder.ASC));
+                    byte sepByte = SchemaUtil.getSeparatorByte(rowKeyOrderOptimizable, ptr.getLength() == 0, isIndexColumnDesc ? SortOrder.DESC : SortOrder.ASC);
+                    output.writeByte(sepByte);
+                    trailingVariableWitdthColumnNum++;
+                } else {
+                    trailingVariableWitdthColumnNum = 0;
                 }
             }
-            int length = stream.size();
-            int minLength = length - maxTrailingNulls;
             byte[] indexRowKey = stream.getBuffer();
             // Remove trailing nulls
-            while (length > minLength && indexRowKey[length-1] == QueryConstants.SEPARATOR_BYTE) {
+            int length = stream.size();
+            int minLength = length - maxTrailingNulls;
+            while (trailingVariableWitdthColumnNum > 0 && length > minLength && indexRowKey[length-1] == QueryConstants.SEPARATOR_BYTE) {
```

Comment:
Could you please add a comment explaining why we need both maxTrailingNulls and trailingVariableWidthColumnNum?

Reply:
The existing code does not eliminate the separator if the data type is not nullable. It not clear why. The actual bug is in the calculation of maxTrailingNulls with view indexes.  So, in order not to impact some other cases, we should keep it here.

Reply:
@gokceni, please add a comment here, e.g., trailingVariableWitdthColumnNum is introduced to fix this issue reported in PHOENIX-6267. The root cause of it is that maxTrailingNulls is not computed correctly for some view index schemas. maxTrailingNulls should be the number of trailing index PK nullable columns. When the actual issue is fixed, trailingVariableWitdthColumnNum can be removed.